File: Nim/perceptron.nim
# nim c -d:release perceptron.nim

import neuron, boolinator

let numInputs = 3
let reps = 600

var inputs = newSeq[bool](numInputs)
var results = newSeq[bool](reps)

var n = newNeuron(numInputs)

for i in 0..<reps:
    for input in 0..<numInputs:
        inputs[input] = nextBool()

    var answer = inputs[0]
    if numInputs > 0:
        for input in 0..<numInputs:
            #answer = answer or inputs[input] # learn to be an OR gate
            answer = answer and inputs[input] # learn to be an AND gate

        answer = not answer # learn to be a NAND gate

    results[i] = n.train(inputs, answer)

for result in results:
    echo result

echo "---------------------------------"
# OR gate
#echo "correct: ", n.forwardBinaryCheck(@[false, false, false], false)
#echo "correct: ", n.forwardBinaryCheck(@[false, false, true], true)
#echo "correct: ", n.forwardBinaryCheck(@[false, true, false], true)
#echo "correct: ", n.forwardBinaryCheck(@[false, true, true], true)
#echo "correct: ", n.forwardBinaryCheck(@[true, false, false], true)
#echo "correct: ", n.forwardBinaryCheck(@[true, false, true], true)
#echo "correct: ", n.forwardBinaryCheck(@[true, true, false], true)
#echo "correct: ", n.forwardBinaryCheck(@[true, true, true], true)

# AND gate
#echo "correct: ", n.forwardBinaryCheck(@[false, false, false], false)
#echo "correct: ", n.forwardBinaryCheck(@[false, false, true], false)
#echo "correct: ", n.forwardBinaryCheck(@[false, true, false], false)
#echo "correct: ", n.forwardBinaryCheck(@[false, true, true], false)
#echo "correct: ", n.forwardBinaryCheck(@[true, false, false], false)
#echo "correct: ", n.forwardBinaryCheck(@[true, false, true], false)
#echo "correct: ", n.forwardBinaryCheck(@[true, true, false], false)
#echo "correct: ", n.forwardBinaryCheck(@[true, true, true], true)

# NAND gate
echo "correct: ", n.forwardBinaryCheck(@[false, false, false], true)
echo "correct: ", n.forwardBinaryCheck(@[false, false, true], true)
echo "correct: ", n.forwardBinaryCheck(@[false, true, false], true)
echo "correct: ", n.forwardBinaryCheck(@[false, true, true], true)
echo "correct: ", n.forwardBinaryCheck(@[true, false, false], true)
echo "correct: ", n.forwardBinaryCheck(@[true, false, true], true)
echo "correct: ", n.forwardBinaryCheck(@[true, true, false], true)
echo "correct: ", n.forwardBinaryCheck(@[true, true, true], false)

File: Rust/src/perceptron.rs
extern crate rand;

mod neuron;

use rand::Rng;
use neuron::Neuron;

fn get_answer(inputs: &Vec<bool>) -> bool {
    let mut answer = inputs[0];
    for input in inputs {
        //answer = answer || *input; // learn to be an OR gate
        answer = answer && *input; // learn to be an AND gate
    }

    answer = !answer; // learn to be a NAND gate

    answer
}

fn main() {
    let num_inputs: u32 = 3;
    let reps = 600;

    let mut inputs: Vec<bool> = Vec::new();
    let mut results: Vec<bool> = Vec::new();

    let mut n = Neuron::new(num_inputs);
    let mut rng = rand::thread_rng();

    for _ in 0..reps {
        for _ in 0..num_inputs {
            inputs.push(rng.gen::<bool>());
        }

        let answer = get_answer(&inputs);
        results.push(n.train(&inputs, &answer));
        inputs.clear();
    }

    for result in results {
        println!("{}", result);
    }

    println!("---------------------------------");
    // OR gate
    //println!("correct: {}", n.forward_binary_check(&vec![false, false, false], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![false, false, true], &true));
    //println!("correct: {}", n.forward_binary_check(&vec![false, true, false], &true));
    //println!("correct: {}", n.forward_binary_check(&vec![false, true, true], &true));
    //println!("correct: {}", n.forward_binary_check(&vec![true, false, false], &true));
    //println!("correct: {}", n.forward_binary_check(&vec![true, false, true], &true));
    //println!("correct: {}", n.forward_binary_check(&vec![true, true, false], &true));
    //println!("correct: {}", n.forward_binary_check(&vec![true, true, true], &true));

    // AND gate
    //println!("correct: {}", n.forward_binary_check(&vec![false, false, false], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![false, false, true], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![false, true, false], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![false, true, true], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![true, false, false], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![true, false, true], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![true, true, false], &false));
    //println!("correct: {}", n.forward_binary_check(&vec![true, true, true], &true));

    // NAND gate
    println!("correct: {}", n.forward_binary_check(&vec![false, false, false], &true));
    println!("correct: {}", n.forward_binary_check(&vec![false, false, true], &true));
    println!("correct: {}", n.forward_binary_check(&vec![false, true, false], &true));
    println!("correct: {}", n.forward_binary_check(&vec![false, true, true], &true));
    println!("correct: {}", n.forward_binary_check(&vec![true, false, false], &true));
    println!("correct: {}", n.forward_binary_check(&vec![true, false, true], &true));
    println!("correct: {}", n.forward_binary_check(&vec![true, true, false], &true));
    println!("correct: {}", n.forward_binary_check(&vec![true, true, true], &false));
}